### PR TITLE
Table ported to react-virtualized

### LINKED ISSUE #664: fixed data table performance
both with the web inspector open and just generally the data table seems to have a fair bit of lag while scrolling (especially horizontally). i haven't been able to pinpoint any commonalities that lead to it, but it feels like it chugs a bit.

### LINKED ISSUE #3751: Tables with a large number of columns can't be displayed as raw data
When displaying a table with a large number of columns (201 in this case), I get an error

screen shot 2016-11-10 at 3 29 55 pm

To replicate --  look at the `Raw Data` of the "Sales Flat Order" of the Magento sample database.

almost certainly related to #664

### PR DESCRIPTION
**Pull request by tlrobinson**
Resolves #664
Resolves #3751

### CODE REVIEW

**Comment by salsakran** [comment]
#3205 is probably also fixed by this.

**Comment by tlrobinson** [comment]
Actually I think that's the same one that we discovered was already fixed. Same as #3687 #3407 and #3751.

**Comment by salsakran** [comment]
can this be merged independently of entity-names? I'd love to get this merged in sooner rather than later

**Comment by tlrobinson** [comment]
I can split it out, shouldn't be too hard.

**Comment by tlrobinson** [comment]
Superseded by #3782

**Comment by tlrobinson** [comment]
Ok I swapped the entity-names branch and this one around so #3723 now builds off of this branch and this can be merged first.

**Comment by tlrobinson** [comment]
I think we're going to need to re-think the way we auto-size columns since we rely on all the columns being rendered but `react-virtualized` renders only the visible ones.

`react-virtualized` has `CellMeasurer`, which I think is for exactly this purpose https://bvaughn.github.io/react-virtualized/#/components/CellMeasurer

**Comment by tlrobinson** [comment]
Ok I changed the way we calculate column widths to do something similar to `CellMeasurer`, basically rendering the first 10 non-empty cells for each column and taking the largest width. This might also be a bit faster since we're not doing two full renders of the table.

**Comment by tlrobinson** [comment]
@kdoh I've worked a bit on the performance. Could you give this another try?

I think vertical scrolling is quite a bit better. Horizontal scrolling isn't ideal, but I'm not sure how much better we can do with `react-virtualized` due to https://github.com/bvaughn/react-virtualized/issues/291

If you want to try the canvas hack you can uncomment this line: https://github.com/metabase/metabase/pull/3752/files#diff-16fac5725b5bb3c4dd5d85421175d35cR321 It makes scrolling less responsive but fixes the lag and flicker issues (in Chrome, not Safari, haven't tested others)

**Comment by tlrobinson** [comment]
This should be ready to go. @metabase/core-developers plz review

File: frontend/src/metabase/visualizations/PieChart.jsx [hunk 1]
```diff
@@ -100,7 +100,7 @@ export default class PieChart extends Component {
         }
 
         let legendTitles = slices.map(slice => [
-            slice.key === "Other" ? slice.key : formatDimension(slice.key, false),
+            slice.key === "Other" ? slice.key : formatDimension(slice.key, true),
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/PieChart.jsx]: Is this part of the table port?

**Review by salsakran — COMMENTED** [review]
I'm not clear on what parts of this are the port and what parts had renamings?

File: frontend/src/metabase/visualizations/PieChart.jsx (hunk 1 above)
Comment (tlrobinson) [inline frontend/src/metabase/visualizations/PieChart.jsx]: No but stuff got shuffled around when I took the entity name stuff out so there might be some random things in here

**Comment by bvaughn** [comment]
Huzzah!

**Pull request merged**